core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -1208,6 +1208,211 @@ private void replaceSubQueries(
+          || valueList.accept(new SqlIdentifierFinder())) {
+        subQuery.expr =
+            convertInToOr(
+                bb,
+                leftKeys,
+                valueList,
+                (SqlInOperator) call.getOperator());
+        return;
+      }
+
+      // Otherwise, let convertExists translate
+      // values list into an inline table for the
+      // reference to Q below.
+    }
+
+    // Project out the search columns from the left side
+
+    // Q1:
+    // "select from emp where emp.deptno in (select col1 from T)"
+    //
+    // is converted to
+    //
+    // "select from
+    //   emp inner join (select distinct col1 from T)) q
+    //   on emp.deptno = q.col1
+    //
+    // Q2:
+    // "select from emp where emp.deptno not in (Q)"
+    //
+    // is converted to
+    //
+    // "select from
+    //   emp left outer join (select distinct col1, TRUE from T) q
+    //   on emp.deptno = q.col1
+    //   where emp.deptno <> null
+    //         and q.indicator <> TRUE"
+    //
+    // Note: Sub-query can be used as SqlUpdate#condition like below:
+    //
+    //   UPDATE emp
+    //   SET empno = 1 WHERE emp.empno IN (
+    //     SELECT emp.empno FROM emp WHERE emp.empno = 2)
+    //
+    // In such case, when converting SqlUpdate#condition, bb.root is null
+    // and it makes no sense to do the sub-query substitution.
+    // However, there are other situations where bb.root is null (specifically the ON condition
+    // of a join) which must be handled.
+
+    final RelDataType targetRowType =
+        SqlTypeUtil.promoteToRowType(typeFactory,
+            validator().getValidatedNodeType(leftKeyNode), null);
+
+    if (bb.root == null) {
+      // Specific handling for IN in the ON clause of a join
+      // (Other sub queries are TODO: https://bodo.atlassian.net/browse/BE-4307)
+      if (bb.inputs != null && bb.inputs.size() == 2 && call.getOperator().kind == SqlKind.IN) {
+
+        //Need requireNonNull wrapper, the checker doesn't see the bb.inputs != null check
+        int outputStartingIdx = requireNonNull(bb.inputs).get(0).getRowType().getFieldCount()
+            + requireNonNull(bb.inputs).get(1).getRowType().getFieldCount();
+
+        // NOTE: this works for non-scalar sub queries, despite the name
+        converted = convertExists(query, RelOptUtil.SubQueryType.SCALAR,
+            subQuery.logic, true, null);
+
+        LogicalAggregate convertedWithDistinct =
+            LogicalAggregate.create(converted.r,
+                ImmutableList.of(),
+                ImmutableBitSet.of(
+                    Util.transform(converted.r.getRowType().getFieldList(),
+                    RelDataTypeField::getIndex)),
+                null,
+                ImmutableList.of()
+                );
+
+
+        // NOTE1: JoinRelType is set to LEFT for the other sub queries that are handled via
+        // joining. I'm not certain why JoinRelType is set to LEFT. To my understanding,
```

Comment:
Could you elaborate on this? What does "other sub queries" mean?

Reply:
added a comment